scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,874 @@
+        ])
+        filtered_tests = fnmatch.filter(all_tests, self.test_pattern)
+        for method_name in filtered_tests:
+            test_fn = getattr(self, 'test_' + method_name)
+            title: str = test_fn.__name__[5:]
+            print(f"{SGR_TEST_BEGIN}Testing {title} ...{SGR_RESET}")
+            try:
+                with JsonRpcProcess(self.solc_path, ["--lsp"], trace_io=self.trace_io) as solc:
+                    test_fn(solc)
+                    self.test_counter.passed += 1
+            except ExpectationFailed as e:
+                self.test_counter.failed += 1
+                print(e)
+                print(traceback.format_exc())
+            except Exception as e: # pragma pylint: disable=broad-except
+                self.test_counter.failed += 1
+                print(f"Unhandled exception {e.__class__.__name__} caught: {e}")
+                print(traceback.format_exc())
+
+        print(
+            f"\n{SGR_NOTICE}Summary:{SGR_RESET}\n\n"
+            f"  Test cases: {self.test_counter.passed} passed, {self.test_counter.failed} failed\n"
+            f"  Assertions: {self.assertion_counter.passed} passed, {self.assertion_counter.failed} failed\n"
+        )
+
+        return min(max(self.test_counter.failed, self.assertion_counter.failed), 127)
+
+    def setup_lsp(self, lsp: JsonRpcProcess, expose_project_root=True):
+        """
+        Prepares the solc LSP server by calling `initialize`,
+        and `initialized` methods.
+        """
+        params = {
+            'processId': None,
+            'rootUri': self.project_root_uri,
+            'trace': 'off',
+            'initializationOptions': {},
+            'capabilities': {
+                'textDocument': {
+                    'publishDiagnostics': {'relatedInformation': True}
+                },
+                'workspace': {
+                    'applyEdit': True,
+                    'configuration': True,
+                    'didChangeConfiguration': {'dynamicRegistration': True},
+                    'workspaceEdit': {'documentChanges': True},
+                    'workspaceFolders': True
+                }
+            }
+        }
+        if expose_project_root == False:
+            params['rootUri'] = None
+        lsp.call_method('initialize', params)
+        lsp.send_notification('initialized')
+
+    # {{{ helpers
+    def get_test_file_path(self, test_case_name):
+        return f"{self.project_root_dir}/{test_case_name}.sol"
+
+    def get_test_file_uri(self, test_case_name):
+        return "file://" + self.get_test_file_path(test_case_name)
+
+    def get_test_file_contents(self, test_case_name):
+        """
+        Reads the file contents from disc for a given test case.
+        The `test_case_name` will be the basename of the file
+        in the test path (test/libsolidity/lsp).
+        """
+        with open(self.get_test_file_path(test_case_name), mode="r", encoding="utf-8") as f:
```

Comment:
This will replace all newlines in the file with `\n`. You need `newline=''` if you want to preserve the original newlines:
```suggestion
        with open(self.get_test_file_path(test_case_name), mode="r", encoding="utf-8", newline='') as f:
```